kitsune/notifications/tasks.py

```diff
@@ -37,26 +38,56 @@ def _full_ct_query(action, actor_only=None):
     return query
 
 
-def _send_simple_push(endpoint, version):
+def _send_simple_push(endpoint, version, max_retries=3, _retry_count=0):
     """
     Hit a simple push endpoint to send a notification to a user.
 
-    Handles and record any HTTP errors.
+    Handles and record any HTTP errors. May retry up to ``max_retries``
+    times by recursing.
+
+    This function tries hard to handle any potential errors, so it may
+    be used in a loop that iterates over many actions to send, without
+    the loop needing to contain error handling logic.
+
+    @param endpoint: The url to PUT to.
+    @param version: The version to include in the push. Should be an integer
+        greater than the version used every other time this endpoint has
+        been called. Timestamps and DB auto increment fields work well.
+    @param max_retries: The maximum number of times to try again.
     """
+
     try:
         r = requests.put(endpoint, 'version={}'.format(version))
-        # If something does wrong, the SimplePush server will give back
-        # json encoded error messages.
-        if r.status_code != 200:
-            logger.error('SimplePush error: %s %s', r.status_code, r.json())
     except RequestException as e:
-        # This will go to Sentry.
-        logger.error('SimplePush PUT failed: %s', e)
+        # This is something like connection error, not a server error.
+        if _retry_count < max_retries:
+            return _send_simple_push(endpoint, version, max_retries, _retry_count + 1)
+        else:
+            logger.error('SimplePush PUT failed: %s', e)
+            return
+
+    # If something does wrong, the SimplePush server should give back json encoded error messages.
+    if r.status_code >= 400:
+        try:
+            data = r.json()
+        except simplejson.scanner.JSONDecodeError:
+            logger.error('SimplePush error (also not JSON?!): %s %s', r.status_code, r.text)
+            return
+
+        if r.status_code == 503 and data['errno'] == 202 and _retry_count < max_retries:
+            return _send_simple_push(endpoint, version, max_retries, _retry_count + 1)
+        else:
+            logger.error('SimplePush error: %s %s', r.status_code, r.json())
 
 
-@task(ignore_result=True)
-def add_notification_for_action(action_id):
-    action = Action.objects.get(id=action_id)
+@task(bind=True)
+def add_notification_for_action(self, action_id):
+    try:
+        action = Action.objects.get(id=action_id)
+    except Action.DoesNotExist as exc:
+        # Maybe database replication is lagging a bit, retry after a few seconds.
+        raise self.retry(exc=exc, max_retries=1, countdown=30)
```

Comment:
For context: This task is run as part of a post-save Django event. We often get errors like [this one](https://errormill.mozilla.org/support/sumo-stage/group/397026/) ([pastebin](https://pastebin.mozilla.org/8850590)).

My theory is that sometimes celery events get processed faster than our database gets replicated. The object *just* got saved. This task is running on a different server than the one that the post-save event happened on, so there won't be any DB pinning. It is likely that the read-DB and the write-DB haven't synced up yet. In that case, waiting a few seconds and trying again would fix it. "A few seconds" could be more accurately described as "whatever our current DB replication lag is". I hope it is less than 30 seconds.

This is happened about 600 times on stage, so it isn't a non-issue. I deployed this PR to stage for testing by BuddyUp people, and those errors stopped happening, so I think my thinking above is correct.